lib/aws/AwsSigv4Signer.js

```diff
@@ -0,0 +1,124 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+'use strict';
+const Connection = require('../Connection');
+const Transport = require('../Transport');
+const aws4 = require('aws4');
+const AwsSigv4SignerError = require('./errors');
+
+function AwsSigv4Signer(opts) {
+  const credentialsState = {
+    credentials: null,
+  };
+  if (opts && (!opts.region || opts.region === null || opts.region === '')) {
+    throw new AwsSigv4SignerError('Region cannot be empty');
+  }
+  if (opts && typeof opts.getCredentials !== 'function') {
+    throw new AwsSigv4SignerError('getCredentials function is required');
+  }
+
+  function buildSignedRequestObject(request = {}) {
+    request.service = 'es';
+    request.region = opts.region;
+    request.headers = request.headers || {};
+    request.headers['host'] = request.hostname;
+    return aws4.sign(request, credentialsState.credentials);
+  }
+
+  class AwsSigv4SignerConnection extends Connection {
+    buildRequestObject(params) {
+      const request = super.buildRequestObject(params);
+      return buildSignedRequestObject(request);
+    }
+  }
+
+  class AwsSigv4SignerTransport extends Transport {
+    request(params, options = {}, callback = undefined) {
+      // options is optional so if options is a function, it's the callback.
+      if (typeof options === 'function') {
+        callback = options;
+        options = {};
+      }
+
+      const currentCredentials = credentialsState.credentials;
+      let expired = false;
+      if (!currentCredentials) {
+        // Credentials haven't been acquired yet.
+        expired = true;
+      }
+      // AWS SDK V2, needsRefresh should be available.
+      else if (typeof currentCredentials.needsRefresh === 'function') {
+        expired = currentCredentials.needsRefresh();
+      }
+      // AWS SDK V2, alternative to needsRefresh.
+      else if (currentCredentials.expired === true) {
+        expired = true;
+      }
+      // AWS SDK V2, alternative to needsRefresh and expired.
+      else if (currentCredentials.expireTime && currentCredentials.expireTime < new Date()) {
```

Comment:
< or <=?

Reply:
I think it's better to be strict in that case for the following reasons:
-previous library would renew the credentials on every requests as now we are refreshing the credentials only when needed
- this condition will probably never been triggered, issue is with the AWS.Credentials object definition that gives 3 different way to test if a Credentials object is expired
- mainly because under heavy load it's possible that between the time we check the request and the time the request is made the Credentials would be expired, so although that condition seems strict, it isn't.

I think it's good like this, but if it's blocker, maybe we can think of exposing an option that gives that choice to the user

something like:
```
new Client(
  ...AwsSigV4({
    ...,
    timeMarginBeforeRenewalMs: 1000 * 60
  })
)
// then that condition would be something like
else if (currentCredentials.expireTime && currentCredentials.expireTime < new Date(Date.now() - (opts.timeMarginBeforeRenewalMs || 0)) {
```
fwiw, I'm good with the current implementation until proven it's an issue, some feedback from users would help to find out.

Reply:
I don't think we should bother with this unless we have some feedback from aws sdk V2 users.
From the aws docs this condition will not be triggered. But because the Credentials object gives those 3 props, for the sake of following the specs it makes sense to test all possible properties.
```
      // AWS SDK V2, needsRefresh should be available.
      else if (typeof currentCredentials.needsRefresh === 'function') {
        expired = currentCredentials.needsRefresh();
      }
      // AWS SDK V2, alternative to needsRefresh.
      else if (currentCredentials.expired === true) {
        expired = true;
      }
      // AWS SDK V2, alternative to needsRefresh and expired.
      else if (currentCredentials.expireTime && currentCredentials.expireTime < new Date()) {
        expired = true;
      }
```
If the Credentials object is consistent in aws sdk v2, `Credentials.needsRefresh()` will always be used.

that point you made and my answer is still relevant for aws sdk v3, in v3 the Credentials object expose only `Credentials.expiration` to test if the credentials are expired.
With a Credentials object from aws sdk V3, this is the only condition that will execute
```
       // AWS SDK V3, Credentials.expiration is a Date object
      else if (currentCredentials.expiration && currentCredentials.expiration < new Date()) {
        expired = true;
      }
```

fwiw, this is the full block that check if the credentials have expired (or haven't been acquired yet)

```
      const currentCredentials = credentialsState.credentials;
      let expired = false;
      if (!currentCredentials) {
        // Credentials haven't been acquired yet.
        expired = true;
      }
      // AWS SDK V2, needsRefresh should be available.
      else if (typeof currentCredentials.needsRefresh === 'function') {
        expired = currentCredentials.needsRefresh();
      }
      // AWS SDK V2, alternative to needsRefresh.
      else if (currentCredentials.expired === true) {
        expired = true;
      }
      // AWS SDK V2, alternative to needsRefresh and expired.
      else if (currentCredentials.expireTime && currentCredentials.expireTime < new Date()) {
        expired = true;
      }
      // AWS SDK V3, Credentials.expiration is a Date object
      else if (currentCredentials.expiration && currentCredentials.expiration < new Date()) {
        expired = true;
      }
```

Reply:
This explanation looks good to me. Looks like the built-in functions from the library can handle this.

Reply:
👍